src/ucp/stream/stream_recv.c

```diff
@@ -268,6 +283,17 @@ UCS_PROFILE_FUNC(ucs_status_ptr_t, ucp_stream_recv_nb,
     ucs_assert(req->recv.state.offset >  0);
     ucs_assert(req->recv.state.offset <= req->recv.length);
     *length = req->recv.state.offset;
+
+    if (ucs_unlikely(req != &tmp_req)) {
+        if (ucp_request_can_complete_stream_recv(req)) {
+            ucp_request_put(req);
```

Comment:
can't we know up-front (e.g during line 253) if the request will complete? so we can use on-stack request instead of mpool get/put

Reply:
I used total data counter for that but you declined this :) now, i do prediction only for user buffer length < 1 rdesc. btw, it would allow to avoid 1 modulo operation.